---
title: "Abu Dhabi RNA-seq project: analysis of pilot data set"
author: "Katie Burnham"
date: '`r Sys.Date()`'
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE)

library(kburnhamfunctions)
library(gdata)
# install.packages("Vennerable", repos = "http://R-Forge.R-project.org")

# Vennerable needs graph and RBGL, from bioconductor repository
# source("http://bioconductor.org/biocLite.R")
# biocLite("graph")
# biocLite("RBGL")

# sample size
library(ssizeRNA)

# Those packages also need reshape and gtools
library(reshape)
library(gtools)
library(Vennerable)

# Differential expression analysis
library(genefilter)
library(DESeq2)

# plotting figures
library(dplyr)
library(ggplot2)
library(ggrepel)
library(gridExtra)

# normalisation
library(vsn)

# heatmap
library(pheatmap)
library(RColorBrewer)

# parallel computing
library(BiocParallel)
register(SnowParam(4))

# look up gene ids
library(ensembldb)
library(EnsDb.Hsapiens.v75)

# pathway analysis
library(XGR)
```

This documents summarises analysis of pilot RNA-seq data from the Abu Dhabi 
cohort metformin response experiment. This comprises samples from 19 individuals 
(9 controls, 10 with type 2 diabetes). NB - patient 115 has been updated to T2D 
based on full clinical information although listed as a control initially).
7 of these have follow up samples taken at a repeat visit 3 months later. For 
each individual and time point, there are 8 different sample types:

- Paxgene

- PBMCs

- CD8

- CD14

- CD14_PBS

- CD14_LPS

- CD14_MET

- CD14_MET_LPS

This gives a total of 208 samples with sequencing data.

**Data locations:**

Fastq files: /well/jknight/Kate/RNAseq/P160233-REX

Sample information in the README file

Clinical information looked up from 
/groups/jknight/Group_members_personal_data_storage/Lawrence/Abu Dabi Patient medical data/Collated Family information.xlsm

Mapping of the sequencing data to the human transcriptome was done by Core 
Bioinformatics using hisat against the human_glk_v37.ERCC reference (Options 
used: Paired reads, deduplicated with Picard, no multiple matches) and features
counted using Subread.

###  Input Data

Full count matrix: /well/jknight/Kate/RNASeq/P160233-REX/Pilot-count-data.txt

The feature count matrix prepared using the [CountMatrix.R script](CountMatrix.R) 
is read in, along with sample information. The value in the i-th row and the j-th
column of the matrix tells how many reads (or fragments, for paired-end RNA-seq) 
have been assigned to gene i in sample j. 

These are not pre-normalized for sequencing depth/library size, as DESeq2's
statistical model is most powerful when applied to un-normalized counts, and is
designed to account for library size differences internally. 

Check that the sample IDs match:

```{r data}
counts <- read.delim("U:/Abu-Dhabi/RNASeq/PilotData/PilotRNA-hisat-count-data.txt")
full.s.info <- read.delim("U:/Abu-Dhabi/RNASeq/PilotData/PilotSampleInfoExpanded.txt",
                          stringsAsFactors=F, row.names=1)
full.s.info$PatientNumber <- as.character(full.s.info$PatientNumber)

update <- read.xls("N:/jknight/Group_members_personal_data_storage/Lawrence/Abu Dabi Patient medical data/FG patient data including 2000 samples (submitted) (updated with diabetes status 18-09-2018).xlsx")

full.s.info$Disease <- update$Status.for.transcriptomics[match(
  paste0("FG", full.s.info$PatientNumber),
  update$Sample.name.on.plate.or.Anonymization.Number)]
full.s.info <- droplevels(full.s.info)
table(full.s.info$Disease)
metrics <- read.delim("U:/Abu-Dhabi/RNASeq/PilotData/Dedup_QC_metric.txt")
full.metrics <- read.delim("U:/Abu-Dhabi/RNASeq/PilotData/Full_QC_metric.txt")

# check order of samples
all(colnames(counts) == rownames(full.s.info))
```

The mapping quality control metrics are explored to determine the quality of the
sequencing data and exclude poor quality samples that would distort downstream
analysis.

```{r mapping-QC}
metrics$SampleType <- full.s.info$SampleType[match(metrics$Sample, 
                                                   full.s.info$SampleId)]
metrics$Disease <- full.s.info$Disease[match(metrics$Sample, 
                                             full.s.info$SampleId)]
metrics$Repeat <- full.s.info$Repeat.Sample[match(metrics$Sample, 
                                             full.s.info$SampleId)]
metrics$Individual <- full.s.info$SampleNumber[match(metrics$Sample,
                                                     full.s.info$SampleId)]

ggplot(metrics, aes(SampleType, Total.Purity.Filtered.Reads.Sequenced)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ylab("Total (purity filtered) reads sequenced)") +
  xlab("Sample Type") +
  ggtitle("Total Reads", 
          subtitle="The total number of reads sequenced per sample is sufficient for all samples\nand doesn't depend on sample type or timepoint")

ggplot(metrics, aes(SampleType, Mapped)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ggtitle("Mapped Reads", "The number of mapped reads is significantly lower for a subset of samples - possible problem") +
  ylab("Number of reads mapped to human genome") +
  xlab("Sample Type")

ggplot(metrics, aes(Total.Purity.Filtered.Reads.Sequenced, Mapped,
                    colour=SampleType)) +
  geom_point() +
  theme_bw() +
  ggtitle("Mapped Reads vs Total Reads Sequenced", 
          "This highlights four samples (4 x R115 cultured CD14) where a large number of sequenced reads were not mapped \nAs these are all from the same patient and time point this seems to be a sample issue") +
  ylab("Number of reads mapped to human genome") +
  xlab("Total (purity filtered) reads sequenced")

ggplot(metrics, aes(SampleType, Mapping.Rate)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ggtitle("Mapping Rate",
          "Highlighting samples with very low mapping rates (<40%) - \nthe 4 samples from R115, R86 CD14_LPS, and 119 CD14_MET") +
  ylab("Percentage of sequenced reads mapped to human genome") +
  xlab("Sample Type") +
  geom_hline(yintercept=0.4)

ggplot(metrics, aes(SampleType, Base.Mismatch.Rate)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ggtitle("Base Mismatch Rate", "Low across all samples, 94 CD8 outlying") +
  ylab("Base mismatch rate") +
  xlab("Sample Type")

ggplot(metrics, aes(SampleType, rRNA.rate)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ggtitle("rRNA Rate", 
          "Low across all - interestingly higher for the cultured CD14s except for the 4 poor quality samples") +
  ylab("rRNA rate") +
  xlab("Sample Type")

ggplot(metrics, aes(SampleType, Intragenic.Rate)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ggtitle("Intragenic Rate",
          "Percentage of reads mapping within genes is generally good - outlying CD14_LPS sample is R86") +
  ylab("Intragenic rate") +
  xlab("Sample Type")

ggplot(metrics, aes(SampleType, Genes.Detected)) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(aes(shape=Repeat, colour=Disease)) +
  theme_bw() +
  ggtitle("Genes Detected with at least 5 reads",
          "Samples with fewest unique genes detected are 4xR115, R86 CD14_LPS, 119 and 100 CD14_MET") +
  ylab("Number of unique genes/transcripts with at least 5 reads mapped") +
  xlab("Sample Type")

sample.type <- as.factor(full.s.info$SampleType)
plot(density(log(counts[, 1]+1)), main="Distribution of raw intensities",
     col=sample.type[1], ylim=c(0, 5.5), xlim=c(-0.8, 3),
     cex.main=0.8)
for (i in 2:ncol(counts)){
  lines(density(log(counts[, i]+1)), col=sample.type[i])
}
legend("topright", legend=unique(sample.type),
       col=c(1:8), lty=1, cex=0.5)
```

Based on these plots, the samples with less than 40% of reads mapped were 
excluded. These are:

```{r to-remove}
to.rm <- as.character(metrics$Sample[metrics$Mapping.Rate < 0.4])
to.rm <- which(full.s.info$SampleId %in% to.rm)
print(full.s.info$SampleNameFull[to.rm])
```

# Differential expression analysis

Explanatory text and the protocol used are largely taken from
[http://www.bioconductor.org/help/workflows/rnaseqGene/](http://www.bioconductor.org/help/workflows/rnaseqGene/), 
'Count-based differential expression analysis of RNA sequencing data using R and
Bioconductor - Anders et al', and the DESeq2 manual.

DESeq2 requires the data to have an associated design formula expressing the 
variables used in modelling. This is used to estimate the dispersions and log2 
FC, and can be changed later if all the DE steps are subsequently repeated. These
data are therefore converted into a DESeq2 dataset with the sample type and 
disease status included in the model to begin with.

NB - under default settings, the final term in the model is the variable of 
interest, with the first level coded as the control group.

```{r convert-to-DESeq2DataSet, warning=FALSE}
counts.rm <- counts[, -to.rm]
full.s.info.rm <- full.s.info[-to.rm, ]

dds <- DESeqDataSetFromMatrix(countData=counts.rm,
                              colData=full.s.info.rm,
                              design= ~ SampleType + Disease)
dds$Disease <- relevel(dds$Disease, ref = "Control")
dds

sample.type <- as.factor(full.s.info.rm$SampleType)
plot(density(log(counts[, 1]+1)), main="Distribution of raw intensities",
     col=sample.type[1], ylim=c(0, 5.5), xlim=c(-0.8, 3),
     cex.main=0.8)
for (i in 2:ncol(counts)){
  lines(density(log(counts[, i]+1)), col=sample.type[i])
}
legend("topright", legend=unique(sample.type), 
       col=c(1:8), lty=1, cex=0.5)
```

### Pre-filtering

We perform a minimal pre-filtering to keep only genes that have at least 10 reads
in 7 or more samples (the number of samples in the smallest biological subgroup).
Note that more strict filtering is applied to increase power via independent 
filtering on the mean of normalized counts within the results function.

The genes identified in this data set may be classified as follows:

```{r pre-filtering}
# filter to keep genes with count >10 in at least 7 samples (the smallest biologically relevant group, 7 in these data)
k <- kOverA(k=7, A=10)
ffun <- filterfun(k)
wh1 <- genefilter(counts.rm, ffun)
sum(wh1) # check how many
filt <- wh1 >=1
dds <- dds[filt, ]
dds
```

## Exploratory analysis and visualisation

Initially we perform some exploratory analysis and look at how the samples relate
to each other. Principal component analysis (PCA) takes all the gene expression 
data and reduces it to a smaller number of independent variables (components)
that summarise the variation in the data i.e. global variation in gene expression.

#### Normalisation

Many common statistical methods for exploratory analysis of multidimensional data,
for example clustering and PCA, work best for data that generally has the same
range of variance at different ranges of the mean values (i.e. homoskedastic). For
RNA-seq counts, however, the expected variance increases with the mean. If one
performs PCA directly on a matrix of normalised counts, the resulting plot
typically depends mostly on the genes with highest counts because they show the
largest absolute differences between samples. A simple strategy to avoid this is
to take the logarithm of the normalised count values plus a pseudocount of 1;
however, depending on the choice of pseudocount, now the genes with the lowest
counts will contribute a great deal of noise to the resulting plot, because taking
the logarithm of small counts actually inflates their variance. 

DESeq2 offers two transformations for count data that stabilise the variance
across the mean: the regularized-logarithm transformation or rlog (Love, Huber,
and Anders 2014), and the variance stabilizing transformation (VST) for negative
binomial data with a dispersion-mean trend (Anders and Huber 2010). For genes with
high counts, the rlog and VST will give similar results to the ordinary log2
transformation of normalised counts. For genes with lower counts, however, the
values are shrunken towards the genes’ averages across all samples. The
rlog-transformed or VST data then becomes approximately homoskedastic, and can be
used directly for computing distances between samples, making PCA plots, or as
input to downstream methods which perform best with homoskedastic data.

The rlog tends to work well on small datasets (n <30), sometimes outperforming
VST when there is a large range of sequencing depth across samples (an order of
magnitude difference). The VST is much faster to compute and is less sensitive to
high count outliers than the rlog, so is therefore recommended for large datasets
(hundreds of samples). On this advice, I have used the VST (this also includes 
correction for sequencing depth). The "blind"=TRUE argument means that the
experimental design is not used to estimate global variability in the counts i.e.
more appropriate for QC purposes.

These plots show the relationship between mean expression and sd before and after
normalisation, confirming that the normalisation has been reasonably successful.

```{r vst, fig.height=3}
vsd <- vst(dds, blind=TRUE)

# Compare before and after
p1 <- meanSdPlot(assay(dds), plot=FALSE)
p1 <- p1$gg + theme_bw()
p2 <- meanSdPlot(assay(vsd), plot=FALSE)
p2 <- p2$gg + theme_bw()

grid.arrange(p1, p2, nrow=1)

melt.exprs <- melt(assay(vsd))
melt.exprs$Type <- full.s.info.rm$SampleType[match(melt.exprs$X2, 
                                                full.s.info.rm$SampleId)]
ggplot(melt.exprs, aes(X2, value, colour=Type)) +
  geom_boxplot() +
  theme_bw() +
  facet_wrap(~ Type, nrow=4, scales="free_x") +
  theme(legend.position="bottom") +
  ggtitle("Boxplots of all data", 
          "There are no outlying samples, although it looks like there are some differences by sample type") +
  xlab("Sample")

pcaData <- plotPCA(vsd, intgroup = c("Disease", "SampleType"), 
                   returnData = TRUE, ntop=nrow(dds))
percentVar <- round(100 * attr(pcaData, "percentVar"))
```

PCA is then used to explore the relationship between the samples. This shows that
a large proportion of the variance (`r paste0(percentVar[1], "% variance")`) is
explained by PC1, which separates the naive and treated cells and also the
non-cultured CD14 cells from the other cell types. There are no clear differences
between diabetes and control samples (shown by the ellipses). Repeat samples were
thought to be of poorer quality, but this does not seem to have an impact on the 
clustering here.

```{r pca}
ggplot(pcaData, aes(x = PC1, y = PC2, color = SampleType, shape = Disease)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme_bw() + 
  stat_ellipse(aes(group=Disease))
```

A second PCA restricted to the cultured (and treated) CD14s shows the general
separation between LPS treated and non-LPS treated cells (shown by the ellipses). 
However, there are a few samples that don't fall within treatment clusters, and 
this seems to relate to the patient number e.g. Patient 102 samples are swapped 
round, before being removed for the above QC reasons 115 all clustered together 
in the centre, and 148 are all at the top left. 

```{r pca-treated}
vsd.treat <- vsd[, vsd$SampleType %in% c("CD14_PBS", "CD14_LPS", 
                                         "CD14_MET", "CD14_MET_LPS")]
vsd.treat$SampleType <- droplevels(vsd.treat$SampleType)
pcaData <- plotPCA(vsd.treat, ntop=nrow(vsd.treat),
                   intgroup = c("Disease", "SampleType", "PatientNumber", 
                                "SampleNumber", "Repeat.Sample"), 
                   returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(x=PC1, y=PC2, color=SampleType, shape=Disease)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme_bw() + stat_ellipse(aes(group=SampleType), linetype = 1) +
  geom_text_repel(data=pcaData[pcaData$PatientNumber %in% c(94, 100, 102, 148), ],
                  aes(label=SampleNumber))
```

This can be seen clearly in the line plot of PC1, where the majority of
individuals show the same pattern across treatments but some clearly stand out.
Specifically, patient 94 looks like the CD14_LPS and CD14_PBS samples have been
swapped, patient 148 perhaps did not get stimulated with LPS, and patients 100 
and 102 look like the LPS treated samples were swapped with the non-treated
samples.

```{r pc1}
ggplot(pcaData, aes(x=SampleType, y=PC1, col=PatientNumber, group=SampleNumber, 
                    shape=Repeat.Sample)) +
  geom_line() +
  geom_point() +
  theme_bw() +
  scale_colour_discrete(guide=FALSE) +
  geom_text(data=pcaData[pcaData$PatientNumber %in% c(94, 100, 102, 148), ],
            aes(label=PatientNumber))
```

These patterns are also seen by clustering the distance matrix for all samples,
and for just the cultured CD14s:

```{r corr-matrix-1, fig.height=15, message=FALSE}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotation_row=data.frame(vsd@colData@listData[c("Disease", "SampleType")])
rownames(annotation_row) <- rownames(sampleDistMatrix)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         annotation_row=annotation_row,
         col = colors)
```

```{r corr-matrix-2, fig.height=10, message=FALSE}
sampleDists <- dist(t(assay(vsd.treat)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- vsd.treat$SampleNameFull
colnames(sampleDistMatrix) <- NULL
annotation_row=data.frame(vsd.treat@colData@listData[c("Disease", "SampleType")])
rownames(annotation_row) <- rownames(sampleDistMatrix)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         annotation_row=annotation_row,
         col = colors)
```

All cultured samples from patients with a clear mix up are removed as we cannot
definitively resolve the issue, and given the small sample size they would have 
a significant impact on the analysis.

```{r remove}
to.rm <- as.character(metrics$Sample[metrics$Mapping.Rate < 0.4])
mixup <- as.character(full.s.info$SampleId[
  grep("94_CD14_.*|148_CD14_.*|100_CD14_.*|102_CD14_.*",
       full.s.info$SampleNameFull)])
to.rm <- c(to.rm, mixup)  
to.rm <- which(full.s.info$SampleId %in% to.rm)
counts.rm <- counts[, -to.rm]
full.s.info.rm <- full.s.info[-to.rm, ]
full.s.info.rm <- droplevels(full.s.info.rm)

dds <- DESeqDataSetFromMatrix(countData=counts.rm,
                              colData=full.s.info.rm,
                              design= ~ SampleType + Disease)
dds$Disease <- relevel(dds$Disease, ref = "Control")

k <- kOverA(k=7, A=10)
ffun <- filterfun(k)
wh1 <- genefilter(counts.rm, ffun)
sum(wh1) # check how many
filt <- wh1 >=1
dds <- dds[filt, ]
dds

# New PCA
vsd <- vst(dds, blind=TRUE)
pcaData <- plotPCA(vsd, intgroup = c("Disease", "SampleType"), returnData = TRUE,
                   ntop=nrow(dds))
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(x = PC1, y = PC2, color = SampleType, shape = Disease)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme_bw() +
  stat_ellipse(aes(group=SampleType))
```

The gene symbols are looked up from their ensembl IDs.

```{r genes}
edb <- EnsDb.Hsapiens.v75
## print some informations for this package
edb
## For what organism was the database generated?
organism(edb)
supportedFilters(edb)

TxByGns <- genes(edb, return.type="data.frame")

featureData <- TxByGns[match(rownames(counts(dds)), TxByGns$gene_id), ]

mcols(dds) <- DataFrame(mcols(dds), featureData)
table(mcols(dds)$gene_biotype)
```

## Differential expression analysis

There are many possible comparisons to perform here, and the experimental design 
is complex (different cell types from the same patients and controls and multiple
time points). Contrasts of interest are:

-	LPS response and the overlap with the same in healthy volunteers

-	Modifier effect of metformin on the LPS response (CD14 LPS vs PBS vs MET_LPS)

-	T2D vs control in each cell type

-	2nd visit (post metformin treatment)

- CD14 vs CD8

- PBMC vs Paxgene

- BMI

From the clinical data, it is clear that there are a range of phenotypes. For 
example, control individuals can have a high BMI, and this should be accounted
for in the analysis. There are also control individuals with PCOS who are being 
treated with metformin (see contingency table). When we compare samples from the 
same patient, all patient-specific factors are accounted for. Following the 
methods used in GWAS, I have included sex and BMI in the model when contrasting disease and controls. 

```{r clinical}
ggplot(full.s.info.rm[!duplicated(full.s.info.rm$PatientNumber), ],
       aes(Disease, HbA1C....)) + 
  geom_boxplot() + 
  theme_bw() + 
  geom_point() +
  geom_hline(yintercept=6.5)

table(full.s.info.rm[!duplicated(full.s.info.rm$PatientNumber), ]$Disease, 
      full.s.info.rm[!duplicated(full.s.info.rm$PatientNumber), ]$Metformin.)
```

```{r normalisation}
dds$PatientNumber <- as.factor(dds$PatientNumber)
dds$Sex <- as.factor(dds$Sex)
design(dds) <- ~ PatientNumber + SampleType
```

### LPS response

The CD14 cells treated with LPS are compared to those treated with PBS (controls
for culturing/treatment), taking account of patient in the model and excluding 
the 2nd visit samples. The log2 FCs are moderated using a normal prior
distribution centered on zero and scaled to fit the data. This removes noise 
arising from low count genes.

P values are adjusted for multiple testing using the Bonferroni-Hochberg FDR, 
and a cut-off of p-adj <0.05 used.

P values for specific genes can be NA if all samples have zero counts, if a
sample has an extreme count outlier (assessed through excluding each sample and 
calculating Cook's distance), and if there is a low mean normalised count
(independent filtering). Given we've already filtered out low count genes, this
step may not have a major effect. These genes are removed from the results table.
Independent filtering is performed based on the mean of normalised counts for each
gene. A threshold is found that optimises the number of adjusted p values lower
than the significance level alpha. This optimisation is visualised below and the 
differential expression results summarised:

```{r lps-response, message=FALSE}
# design
lps.dds <- dds[, dds@colData@listData$Repeat.Sample == FALSE]
lps.dds@colData <- droplevels(lps.dds@colData)
#lps.dds$SampleType <- relevel(lps.dds$SampleType, ref = "CD14_PBS")
lps.dds$PatientNumber <- as.factor(lps.dds$PatientNumber)
table(lps.dds$SampleType)

# analysis
lps.dds <- DESeq(lps.dds, parallel=TRUE)

# results
lps.res <- results(lps.dds, parallel=TRUE,
                   contrast=c("SampleType", "CD14_LPS", "CD14_PBS"),
                   independentFiltering=TRUE, alpha=0.05)
lps.res$Gene <- mcols(lps.dds)$gene_name

plot(metadata(lps.res)$filterNumRej,
     type="b", ylab="Number of rejections",
     xlab="Quantiles of filter", pch=16)
lines(metadata(lps.res)$lo.fit, col="red")
abline(v=metadata(lps.res)$filterTheta)

lps.res.LFC <- lfcShrink(lps.dds, 
                         contrast=c("SampleType", "CD14_LPS", "CD14_PBS"),
                         parallel=TRUE)
lps.res$log2FoldChange <- lps.res.LFC$log2FoldChange
lps.res$lfcSE <- lps.res.LFC$lfcSE
lpsresOrdered <- lps.res[order(lps.res$padj), ]
lpsresOrdered <- lpsresOrdered[(complete.cases(lpsresOrdered)), ]
summary(lpsresOrdered)
res05 <- results(lps.dds, alpha=0.05)

# write.table(lpsresOrdered,
            # file="U:/Abu-Dhabi/RNASeq/PilotData/LPSResponse_All_Patients.txt",
            # sep="\t")
```

`r sum(lpsresOrdered$padj < 0.05, na.rm=TRUE)` transcripts are differentially
expressed at FDR <0.05. 

Various plots are used for QC. The MA plot shows that differential expression is
not overly dependent on mean expression levels, and that the fold changes are
centered around 0. 

```{r lps-response-maplot}
# QC plots
r.lfc <- range(lps.res.LFC$log2FoldChange, na.rm=TRUE)
plotMA(lps.res.LFC, ylim=r.lfc, main="MA Plot")
abline(h=c(-log(1.5, 2), log(1.5, 2)), col="dodgerblue", lwd=2)
```

The dispersion estimates are typical, with the final estimates shrunk from the
gene-wise estimates towards the fitted estimates. 

```{r lps-response-dispplot}
plotDispEsts(lps.dds, main="Dispersion plot")
```

There are no outlying count values, and a similar distribution is seen across 
samples for Cook's distance.

```{r lps-response-cooksplot}
boxplot(log10(assays(lps.dds)[["cooks"]]), range=0, las=2, main="Cook's distance")
# No outliers 
```

The p value histogram is as expected, with a uniform distribution for 
non-significant genes and a spike on the left hand side for biologically 
different genes.

```{r lps-response-pvalhist}
hist(lpsresOrdered$pvalue, breaks=100, main="Histogram of filtered p values",
     xlab="P value")
```

The top ten most significantly differentially expressed genes are plotted across
samples. 

```{r lps-response-de-plots, fig.height=10}
# Top DE genes
top.10 <- order(lps.res$padj)[1:10]
de.plot.list <- list()
for(j in 1:10){
  i <- top.10[j]
  d <- plotCounts(lps.dds, gene=i, 
                intgroup=c("SampleType", "Disease", "PatientNumber"),
                returnData=TRUE)
  d <- subset(d, SampleType %in% c("CD14_LPS", "CD14_PBS"))
  gene.name <- as.character(mcols(lps.dds)$gene_name[i])
  if(gene.name == "") gene.name <- as.character(mcols(lps.dds)$gene_id[i])
  de.plot.list[[j]] <- ggplot(d, aes(x=SampleType, y=count, color=Disease,
                                     group=PatientNumber)) + 
    geom_point() + 
    scale_y_log10() +
    theme_bw() +
    geom_line() +
    ylab(gene.name)
}

grid.arrange(grobs=de.plot.list, ncol=2)
```

This volcano plot gives an overview of the results, with some of the most
significantly differentially expressed genes labelled and a positive fold change
indicating higher expression in the LPS treated cells. 

```{r lps-response-vpplot}
lpsresOrdered$sig <- ifelse((lpsresOrdered$padj < 0.05 & 
                            abs(lpsresOrdered$log2FoldChange) >= log(1.5, 2)), 
                          "Sig DE", "Not Sig")
ggplot(data.frame(lpsresOrdered), aes(log2FoldChange, -log10(padj))) +
  geom_point(aes(col=sig)) +
  scale_color_manual(values=c("grey", "red")) +
  ylab("-log10(adjusted p value)") +
  xlab("log2(fold change)") +
  ggtitle("LPS vs PBS CD14") +
  theme_bw() +
  geom_text_repel(data=subset(data.frame(lpsresOrdered), padj<5e-100),
                  aes(label=Gene))
```

Pathway enrichment analysis is performed with XGR, using the lea algorithm to 
account for hierarchical relationships between terms and the GOBP ontology. The
specified background is all genes tested for differential expression after
filtering. The ten terms most significantly enriched in the list of differentially
expressed genes are shown:

```{r lps-xgr, message=FALSE}
RData.location <- "http://galahad.well.ox.ac.uk/bigdata"

data <- unique(as.character(lpsresOrdered$Gene[lpsresOrdered$sig == "Sig DE"]))
background <- unique(as.character(lpsresOrdered$Gene))
eTerm_LPSAll_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                        ontology="GOBP",
                                        ontology.algorithm="lea",
                                        RData.location=RData.location,
                                        verbose=FALSE)
bp_LPSAll_GOBP_lea <- xEnrichBarplot(eTerm_LPSAll_GOBP_lea,
                                   displayBy="fdr", top_num=10)
bp_LPSAll_GOBP_lea
```

### Comparison to healthy volunteers

These results are positively correlated with the CD14 LPS response from the 
Fairfax et al healthy volunteer data (Science 2014). Due to the different
platforms used (microarray vs RNA-seq), for this comparison mean log fold changes
were calculated for each gene and the value for the Fairfax cohort plotted against
that for the Abu Dhabi cohort. The Fairfax data included a 2 hour and a 24 hour
stimulation of CD14s with LPS, so the Abu Dhabi results (6 hour treatment with 
LPS) are compared to both:

```{r lps2-correlation, warning=FALSE}
correlation.plot <- function (limma.1, limma.2, name.1 = "A", name.2 = "B", 
          main.title = "Correlation Plot", 
          key = TRUE, col.1 = "darkgreen", col.2 = "gold", col.both = "blue", 
  labels = FALSE) {
  
  limma.1 <- limma.1[complete.cases(limma.1), ]
  limma.2 <- limma.2[complete.cases(limma.2), ]
  
  limma.2 <- limma.2[which(limma.2$Gene %in% limma.1$Gene), 
    ]
  limma.1 <- limma.1[which(limma.1$Gene %in% limma.2$Gene), 
    ]
  limma.2 <- limma.2[match(limma.1$Gene, limma.2$Gene), 
    ]
  
  plot.range.x <- max(abs(c(limma.1$log2FoldChange)))
  plot.range.y <- max(abs(c(limma.2$log2FoldChange)))
  
  corr.plot.colours <- rep("gray", nrow(limma.1))
  for (i in 1:nrow(limma.1)) {
    if (abs(limma.1$log2FoldChange[i]) > log2(1.5) & 
        limma.1$padj[i] < 0.05 & 
        abs(limma.2$log2FoldChange[i]) > log2(1.5) & 
        limma.2$pad[i] < 0.05) {
      corr.plot.colours[i] <- col.both
    }
    else if ((abs(limma.1$log2FoldChange[i]) > log2(1.5) & 
              limma.1$padj[i] < 0.05) & 
             (abs(limma.2$log2FoldChange[i]) < log2(1.5) | limma.2$padj[i] > 0.05))
      {
      corr.plot.colours[i] <- col.1
    }
    else if ((abs(limma.1$log2FoldChange[i]) < log2(1.5) | 
              limma.1$padj[i] > 0.05) & 
             (abs(limma.2$log2FoldChange[i]) > log2(1.5) & limma.2$padj[i] < 0.05))
      {
      corr.plot.colours[i] <- col.2
    }
    if (limma.1$log2FoldChange[i] * limma.2$log2FoldChange[i] < 0) {
      if (abs(limma.1$log2FoldChange[i]) > log2(1.5) & limma.1$padj[i] < 0.05 &
          abs(limma.2$log2FoldChange[i]) > log2(1.5) & limma.2$padj[i] < 0.05) 
        {
        corr.plot.colours[i] <- "red"
      }
    }
  }
  
  plot(limma.1$log2FoldChange, limma.2$log2FoldChange, 
       xlab = paste("Log2 Fold Change", name.1), 
       ylab = paste("Log2 Fold Change", name.2), 
       main = main.title, 
       xlim = c(-plot.range.x, plot.range.x), 
       ylim = c(-plot.range.y, plot.range.y), 
       col = corr.plot.colours, pch = 20)
  
  text(x = (plot.range.x - 1), y = -(plot.range.y - 1), paste("Pearson's r =", 
    format(cor(limma.1$log2FoldChange, limma.2$log2FoldChange), digits = 3)))
  corr.test <- cor.test(limma.1$log2FoldChange, limma.2$log2FoldChange)
  if (as.numeric(corr.test$p.value) > 0) {
    text(x = (plot.range.x - 1), y = -(plot.range.y - 0.7), 
      paste("p =", format(corr.test$p.value, digits = 3)))
  }
  else {
    text(x = (plot.range.x - 1), y = -(plot.range.y - 0.7), 
      "P < 2.2 x 10-16")
  }
  if (key == TRUE) {
    legend("topleft", c("DE in Both", paste("DE in", name.1), 
      paste("DE in", name.2), "DE in Opposite Directions", 
      "Not DE"), bty = "n", col = c(col.both, col.1, col.2, 
      "red", "gray"), pch = rep(20, 4))
  }
  abline(h = 0)
  abline(v = 0)
  if (labels == TRUE) {
    text(limma.2$Gene, x = limma.1$log2FoldChange, 
         y = limma.2$log2FoldChange + 0.1, cex = 0.8)
  }
}

fairfax <- read.delim("U:/Abu-Dhabi/RNASeq/PilotData/FairfaxScienceLPSDE.txt",
                      stringsAsFactors=FALSE)
fairfax.lps <- fairfax[, c("Symbol", "logFC.LPS2.vs.Naive",
                           "adj.P.Val.LPS2.vs.Naive")]
fairfax.lps <- fairfax.lps[complete.cases(fairfax.lps), ]
fairfax.lps$logFC.LPS2.vs.Naive <- (-fairfax.lps$logFC.LPS2.vs.Naive)

fairfax.lps <- aggregate(fairfax.lps[, 2:3], by=list(fairfax.lps$Symbol), 
                         FUN=mean)
colnames(fairfax.lps) <- c("Gene", "log2FoldChange", "padj")
lps.de <- aggregate(lpsresOrdered[, c(2, 6, 7)], by=list(lpsresOrdered$Gene),
                    FUN=mean)
lps.de$Gene <- NULL
colnames(lps.de)[1] <- "Gene"
lps.de <- lps.de[complete.cases(lps.de), ]
lps.de$Gene <- as.character(lps.de$Gene)
correlation.plot(lps.de, fairfax.lps, "Pilot", "Fairfax 2h", "LPS Response")
```

```{r lps24-correlation}
fairfax.lps <- fairfax[, c("Symbol", "logFC.LPS24.vs.Naive",
                           "adj.P.Val.LPS24h.vs.Naive")]
fairfax.lps <- fairfax.lps[complete.cases(fairfax.lps), ]
fairfax.lps$logFC.LPS24.vs.Naive <- (-fairfax.lps$logFC.LPS24.vs.Naive)

fairfax.lps <- aggregate(fairfax.lps[, 2:3], by=list(fairfax.lps$Symbol), 
                         FUN=mean)
colnames(fairfax.lps) <- c("Gene", "log2FoldChange", "padj")
correlation.plot(lps.de, fairfax.lps, "Pilot", "Fairfax 24h", "LPS Response")
```

#### Disease status and the LPS response

In order to test for an effect of disease status on the LPS response, the sample
information must be recoded (as described in the DESeq2 manual). Briefly, the
model is not full rank when accounting for individual, as each individual is
either in the disease or control group. Individual is recoded within each group,
levels without samples removed from the design matrix, and the LPS response in
disease and controls contrasted to test if the treatment effect is different
across groups. There are no significant genes.

```{r lps-response-disease, message=FALSE}
# remove pre-diabetes
lps.dds <- lps.dds[, -(which(lps.dds$Disease == "Pre-diabetes"))]
lps.dds@colData <- droplevels(lps.dds@colData)
# to account for the effect of disease status
# two disease status groups with distinct individuals
tmp <- cbind(lps.dds$Disease, lps.dds$PatientNumber)
# 10 patients in control, 9 in disease
# update disease to 1:9 and control 1:10
ind.n <- tmp[, 2]
ind.n[which(tmp[, 1] == 1)] <- as.numeric(as.factor(ind.n[which(tmp[, 1] == 1)]))
ind.n[which(tmp[, 1] == 2)] <- as.numeric(as.factor(ind.n[which(tmp[, 1] == 2)]))

lps.dds.dis <- lps.dds
lps.dds.dis$ind.n <- as.factor(as.character(ind.n))

design(lps.dds.dis) <- ~ Disease + Disease:ind.n + Disease:SampleType
m1 <- model.matrix(design(lps.dds.dis), lps.dds.dis@colData)
all.zero <- colSums(m1) == 0
idx <- which(all.zero)
m1 <- m1[, -idx]

# analysis
lps.dds.dis <- DESeq(lps.dds.dis, parallel=TRUE, full=m1, betaPrior=FALSE)

# results
lps.dis.res <- results(lps.dds.dis, parallel=TRUE, 
                   independentFiltering=TRUE, alpha=0.05,
                   contrast=list("DiseaseT2DM.SampleTypeCD14_LPS",
                                 "DiseaseControl.SampleTypeCD14_LPS"))
lps.dis.res$Gene <- mcols(lps.dds.dis)$gene_name
lps.dis.res <- lps.dis.res[(complete.cases(lps.dis.res)), ]
summary(lps.dis.res)
lps.dis.res[lps.dis.res$padj < 0.05, ]
```

I have also analysed the patients and controls separately:

```{r lps-response-controls, message=FALSE, warning=FALSE}
lps.con.dds <- dds[, dds@colData@listData$SampleType %in% c("CD14_PBS",
                                                            "CD14_LPS")]
lps.con.dds <- lps.con.dds[, lps.con.dds@colData@listData$Repeat.Sample == FALSE]
lps.con.dds <- lps.con.dds[, lps.con.dds@colData@listData$Disease == "Control"]
# 20 samples remaining (10 individuals, 2 samples each)
lps.con.dds@colData <- droplevels(lps.con.dds@colData)
lps.con.dds$PatientNumber <- as.factor(lps.con.dds$PatientNumber)
lps.con.dds$SampleType <- relevel(lps.con.dds$SampleType, ref="CD14_PBS")
design(lps.con.dds) <- ~ PatientNumber + SampleType
table(lps.con.dds$SampleType)

# analysis
lps.con.dds <- DESeq(lps.con.dds)

# normalised counts
lps.con.dds.norm <- counts(lps.con.dds, normalized=TRUE)

# results
lps.con.res <- results(lps.con.dds, 
                       independentFiltering=TRUE, alpha=0.05)
lps.con.res$Gene <- mcols(lps.con.dds)$gene_name

plot(metadata(lps.con.res)$filterNumRej,
     type="b", ylab="Number of rejections",
     xlab="Quantiles of filter", pch=16)
lines(metadata(lps.con.res)$lo.fit, col="red")
abline(v=metadata(lps.con.res)$filterTheta)

lps.con.res.LFC <- lfcShrink(lps.con.dds, coef=length(resultsNames(lps.con.dds)))
lps.con.res$log2FoldChange <- lps.con.res.LFC$log2FoldChange
lps.con.res$lfcSE <- lps.con.res.LFC$lfcSE
lpsconresOrdered <- lps.con.res[order(lps.con.res$padj), ]
lpsconresOrdered <- lpsconresOrdered[(complete.cases(lpsconresOrdered)), ]
summary(lpsconresOrdered)

# write.table(lpsconresOrdered,
            # file="U:/Abu-Dhabi/RNASeq/PilotData/LPSResponse_Controls.txt",
            # sep="\t")

# QC plots
r.lfc <- range(lps.con.res.LFC$log2FoldChange, na.rm=TRUE)
plotMA(lps.con.res.LFC, ylim=r.lfc, main="MA Plot")
abline(h=c(-log(1.5, 2), log(1.5, 2)), col="dodgerblue", lwd=2)

plotDispEsts(lps.con.dds, main="Dispersion plot")

boxplot(log10(assays(lps.con.dds)[["cooks"]]), range=0, las=2, 
        main="Cook's distance")

hist(lpsconresOrdered$pvalue, breaks=100, main="P value")

lps.con.de <- as.data.frame(aggregate(lpsconresOrdered[, c(2, 6, 7)],
                                      by=list(lpsconresOrdered$Gene),
                                      FUN=mean))
lps.con.de$Gene <- NULL
colnames(lps.con.de)[1] <- "Gene"
lps.con.de <- lps.con.de[complete.cases(lps.con.de), ]
lps.con.de$Gene <- as.character(lps.con.de$Gene)
correlation.plot(lps.de, lps.con.de, "Pilot", "Pilot - Controls", "LPS Response")
```

```{r lps-response-t2d, warning=FALSE, message=FALSE}
lps.t2d.dds <- dds[, dds@colData@listData$SampleType %in% c("CD14_PBS",
                                                            "CD14_LPS")]
lps.t2d.dds <- lps.t2d.dds[, lps.t2d.dds@colData@listData$Repeat.Sample == FALSE]
lps.t2d.dds <- lps.t2d.dds[, lps.t2d.dds@colData@listData$Disease == "T2DM"]
# 18 samples remaining (9 individuals, 2 samples each)
lps.t2d.dds@colData <- droplevels(lps.t2d.dds@colData)
lps.t2d.dds$SampleType <- relevel(lps.t2d.dds$SampleType, ref="CD14_PBS")
lps.t2d.dds$Disease <- droplevels(lps.t2d.dds$Disease)
lps.t2d.dds$PatientNumber <- as.factor(lps.t2d.dds$PatientNumber)
design(lps.t2d.dds) <- ~ PatientNumber + SampleType
table(lps.t2d.dds$SampleType)

# analysis
lps.t2d.dds <- DESeq(lps.t2d.dds)

# normalised counts
lps.t2d.dds.norm <- counts(lps.t2d.dds, normalized=TRUE)

# results
lps.t2d.res <- results(lps.t2d.dds, 
                       independentFiltering=TRUE, alpha=0.05)
lps.t2d.res$Gene <- mcols(lps.t2d.dds)$gene_name

plot(metadata(lps.t2d.res)$filterNumRej,
     type="b", ylab="Number of rejections",
     xlab="Quantiles of filter", pch=16)
lines(metadata(lps.t2d.res)$lo.fit, col="red")
abline(v=metadata(lps.t2d.res)$filterTheta)

lps.t2d.res.LFC <- lfcShrink(lps.t2d.dds, coef=length(resultsNames(lps.t2d.dds)))
lps.t2d.res$log2FoldChange <- lps.t2d.res.LFC$log2FoldChange
lps.t2d.res$lfcSE <- lps.t2d.res.LFC$lfcSE
lpst2dresOrdered <- lps.t2d.res[order(lps.t2d.res$padj), ]
lpst2dresOrdered <- lpst2dresOrdered[(complete.cases(lpst2dresOrdered)), ]
summary(lpst2dresOrdered)

# write.table(lpst2dresOrdered,
            # file="U:/Abu-Dhabi/RNASeq/PilotData/LPSResponse_T2D.txt",
            # sep="\t")

# QC plots
r.lfc <- range(lps.t2d.res.LFC$log2FoldChange, na.rm=TRUE)
plotMA(lps.t2d.res.LFC, ylim=r.lfc, main="MA plot")
abline(h=c(-log(1.5, 2), log(1.5, 2)), col="dodgerblue", lwd=2)

plotDispEsts(lps.t2d.dds, main="Dispersion plot")

boxplot(log10(assays(lps.t2d.dds)[["cooks"]]), range=0, las=2,
        main="Cook's distance")
# No outliers 

hist(lpst2dresOrdered$pvalue, breaks=100, main="P values")

lps.t2d.de <- as.data.frame(aggregate(lpst2dresOrdered[, c(2, 6, 7)],
                                      by=list(lpst2dresOrdered$Gene),
                                      FUN=mean))
lps.t2d.de$Gene <- NULL
colnames(lps.t2d.de)[1] <- "Gene"
lps.t2d.de <- lps.t2d.de[complete.cases(lps.t2d.de), ]
lps.t2d.de$Gene <- as.character(lps.t2d.de$Gene)
correlation.plot(lps.de, lps.t2d.de, "Pilot", "Pilot - T2D", "LPS Response")
correlation.plot(lps.con.de, lps.t2d.de, 
                 "Controls", "T2D", "LPS Response")
```

Pathway analysis using XGR of LPS response in the different data sets:

```{r lps-xgr-hv, fig.height=12, message=FALSE}
# Fairfax
data <- fairfax$Symbol[fairfax$adj.P.Val.LPS2.vs.Naive < 0.05 &
                         abs(fairfax$logFC.LPS2.vs.Naive) > log(1.5, 2)]
background <- fairfax$Symbol
eTerm_LPS2_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                      ontology="GOBP",
                                      ontology.algorithm="lea",
                                     RData.location=RData.location,
                                     verbose=FALSE)

# Controls
data <- as.character(lpsconresOrdered$Gene[lpsconresOrdered$padj < 0.05 &
                              abs(lpsconresOrdered$log2FoldChange) > log(1.5, 2)])
background <- as.character(lpsconresOrdered$Gene)
eTerm_LPSControl_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                          ontology="GOBP",
                                          ontology.algorithm="lea",
                                          RData.location=RData.location,
                                          verbose=FALSE)

# T2D
data <- as.character(lpst2dresOrdered$Gene[lpst2dresOrdered$padj < 0.05 &
                              abs(lpst2dresOrdered$log2FoldChange) > log(1.5, 2)])
background <- as.character(lpst2dresOrdered$Gene)
eTerm_LPST2D_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                        ontology="GOBP",
                                        ontology.algorithm="lea",
                                        RData.location=RData.location,
                                        verbose=FALSE)

# Compare
list_eTerm <- list(eTerm_LPSAll_GOBP_lea, eTerm_LPSControl_GOBP_lea,
                   eTerm_LPST2D_GOBP_lea)
names(list_eTerm) <- c('Abu Dhabi (All)', 'Controls', "T2D")
xEnrichCompare(list_eTerm, displayBy="fc", FDR.cutoff=0.001) + 
  theme(axis.text.y=element_text(size=10))
```

### Modifier effect of MET

To explore the modifying effect of metformin, we want to contrast the LPS vs PBS
response to the CD14_MET_LPS vs PBS response.

NB might need to remove 119 as no CD14_MET sample

```{r lps-met, message=FALSE, warning=FALSE}
# results
pbs.met.res <- results(lps.dds, 
                       contrast=c("SampleType", "CD14_PBS", "CD14_MET"),
                   independentFiltering=TRUE, alpha=0.05)
pbs.met.res$Gene <- mcols(lps.dds)$gene_name
pbs.met.res.LFC <- lfcShrink(lps.dds, 
                             contrast=c("SampleType", "CD14_PBS", "CD14_MET"))
pbs.met.res$log2FoldChange <- pbs.met.res.LFC$log2FoldChange
pbs.met.res$lfcSE <- pbs.met.res.LFC$lfcSE
pbsmetresOrdered <- pbs.met.res[order(pbs.met.res$padj), ]
pbsmetresOrdered <- pbsmetresOrdered[(complete.cases(pbsmetresOrdered)), ]
print("CD14_PBS vs CD14_MET")
summary(pbsmetresOrdered)

d <- plotCounts(lps.dds, gene="ENSG00000262292", 
                intgroup=c("SampleType", "Disease", "PatientNumber"),
                returnData=TRUE)
d <- subset(d, SampleType %in% c("CD14_MET", "CD14_PBS"))
gene.name <- "RP11-160E2.11"
ggplot(d, aes(x=SampleType, y=count, color=Disease,
                                     group=PatientNumber)) + 
    geom_point(position=position_dodge(width=0.1)) + 
    theme_bw() +
    geom_line() +
    ylab(gene.name)

lps.met.res <- results(lps.dds, 
                   contrast=c("SampleType", "CD14_LPS", "CD14_MET_LPS"),
                   independentFiltering=TRUE, alpha=0.05)
lps.met.res$Gene <- mcols(lps.dds)$gene_name
lps.met.res.LFC <- lfcShrink(lps.dds, 
                             contrast=c("SampleType", "CD14_LPS",
                                        "CD14_MET_LPS"))
lps.met.res$log2FoldChange <- lps.met.res.LFC$log2FoldChange
lps.met.res$lfcSE <- lps.met.res.LFC$lfcSE
lpsmetresOrdered <- lps.met.res[order(lps.met.res$padj), ]
lpsmetresOrdered <- lpsmetresOrdered[(complete.cases(lpsmetresOrdered)), ]
print("CD14_LPS vs CD14_MET_LPS")
summary(lpsmetresOrdered)

pbs.met.res <- results(lps.dds, 
                   contrast=c("SampleType", "CD14_PBS", "CD14_MET_LPS"),
                   independentFiltering=TRUE, alpha=0.05)
pbs.met.res$Gene <- mcols(lps.dds)$gene_name
pbs.met.res.LFC <- lfcShrink(lps.dds, 
                             contrast=c("SampleType", "CD14_MET_LPS",
                                        "CD14_PBS"))
pbs.met.res$log2FoldChange <- pbs.met.res.LFC$log2FoldChange
pbs.met.res$lfcSE <- pbs.met.res.LFC$lfcSE
metresOrdered <- pbs.met.res[order(pbs.met.res$padj), ]
metresOrdered <- metresOrdered[(complete.cases(metresOrdered)), ]
print("CD14_MET_LPS vs CD14_PBS")
summary(metresOrdered)

pbs.met.de <- as.data.frame(aggregate(metresOrdered[, c(2, 6, 7)],
                                      by=list(metresOrdered$Gene),
                                      FUN=mean))
pbs.met.de$Gene <- NULL
colnames(pbs.met.de)[1] <- "Gene"
pbs.met.de <- pbs.met.de[complete.cases(pbs.met.de), ]
pbs.met.de$Gene <- as.character(pbs.met.de$Gene)
correlation.plot(lps.de, pbs.met.de, "LPS Response", "MET_LPS response", 
                 "LPS Response")
```

No genes are differentially expressed between LPS and MET_LPS. One gene is 
differentially expressed between MET and PBS treated cells, but it is not very
convincing.

### T2D vs Controls

Patients and controls are compared for each cell type. The repeat samples are
excluded, so 10 patients are compared to 9 controls. The power to detect a 1.5
fold change, assuming that 1% of genes are DE, is estimated from the data.

```{r disease, message=FALSE}
# design
disease.dds <- lps.dds
# 127 samples remaining (18 individuals, 4-8 samples each)
design(disease.dds) <- ~ BMI + Sex + Disease
table(disease.dds$SampleType)

# analysis
disease.res.list <- list()
disease.plot.list <- list()

#power calc parameters
# proportion of non-DE genes
pi0 <- 0.99
# FDR level to control
fdr <- 0.05
# desired average power
power <- 0.8
# fold change for each gene
fc <- 1.5
# pseudo sample size
m <- 9
# maximum sample size used for power calculations
maxN <- 30

for(i in 1:8){
  cell.type <- levels(disease.dds$SampleType)[i]
  print(cell.type)
  sub.dis.dds <- disease.dds[, disease.dds$SampleType == cell.type]
  sub.dis.dds@colData <- droplevels(sub.dis.dds@colData)
  keep <- rowSums(counts(sub.dis.dds)) >= 10
  sub.dis.dds <- sub.dis.dds[keep,]
  sub.dis.dds
  sub.dis.dds <- DESeq(sub.dis.dds)
  disease.res <- results(sub.dis.dds, 
                   independentFiltering=TRUE, alpha=0.05)
  disease.res$Gene <- mcols(sub.dis.dds)$gene_name
  disease.res.LFC <- lfcShrink(sub.dis.dds, 
                               coef=length(resultsNames(sub.dis.dds)))
  disease.res$log2FoldChange <- disease.res.LFC$log2FoldChange
  disease.res$lfcSE <- disease.res.LFC$lfcSE
  
  # power
  # total number of genes for testing
  G <- nrow(disease.res)
  # average read count for each gene in control group
  mu <- disease.res$baseMean
  # dispersion parameter for each gene
  disp <- dispersions(sub.dis.dds)

  set.seed(19042018)
  size1 <- ssizeRNA_vary(nGenes=G,
                         pi0=pi0,
                         m=m,
                         mu=mu,
                         disp=disp,
                         fc=fc,
                         fdr=fdr,
                         power=power,
                         maxN=maxN)

  check.power(m=9, mu=mu, disp=disp, fc=1.5, sims=10)
  
  diseaseresOrdered <- disease.res[order(disease.res$padj), ]
  diseaseresOrdered <- diseaseresOrdered[(complete.cases(diseaseresOrdered)), ]
  disease.res.list[[i]] <- diseaseresOrdered
  summary(diseaseresOrdered)
  
  diseaseresOrdered$sig <- ifelse((diseaseresOrdered$padj < 0.05), 
                          "Sig DE", "Not Sig")
  disease.plot.list[[i]] <- ggplot(data.frame(diseaseresOrdered),
                                   aes(log2FoldChange, -log10(padj))) +
    geom_point(aes(col=sig)) +
    scale_color_manual(values=c("grey", "red")) +
    theme_bw() +
    ggtitle(cell.type) +
    geom_text_repel(data=subset(data.frame(diseaseresOrdered), sig == "Sig DE"),
                    aes(label=Gene))
}

grid.arrange(grobs=disease.plot.list, ncol=3)
```

The specific genes of interest are plotted:

```{r disease-genes, warning=FALSE}
g <- which(mcols(disease.dds)$gene_name == "RPL41P1")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("RPL41P1") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "C17orf97")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("C17orf97") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "TRAPPC6A")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("TRAPPC6A") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "PDPN")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("PDPN") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "SCD")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("SCD") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "TPRG1")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("TPRG1") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "LYN")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("LYN") +
  facet_wrap(~ SampleType, nrow=2)

g <- which(mcols(disease.dds)$gene_name == "AC008268.2")
d <- plotCounts(disease.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease"),
                returnData=TRUE)
ggplot(d, aes(x=Disease, y=count, color=Disease)) + 
  geom_boxplot() +
  geom_point(size=3) + 
  scale_y_log10() +
  theme_bw() +
  ylab("AC008268.2") +
  facet_wrap(~ SampleType, nrow=2)
```

### Repeat visit

Individuals without a repeat visit are excluded, and changes over time identified
for each cell type. 

NB the repeat samples were poorer quality, which could have a batch effect that
can't be removed.

```{r repeat, message=FALSE}
repeat.inds <- names(which(table(dds$PatientNumber) >= 12))
repeat.dds <- dds[, dds@colData@listData$PatientNumber %in% repeat.inds]
to.rm <- c("86_CD14_LPS", 
           "115_CD14_MET_LPS", "115_CD14_PBS", "115_CD14_MET",
           "115_CD14_LPS")
repeat.dds <- repeat.dds[, !(repeat.dds$SampleNameFull %in% to.rm)]
# 102 samples remaining (7 individuals, 16 samples each)
repeat.dds$PatientNumber <- as.factor(repeat.dds$PatientNumber)
repeat.dds@colData <- droplevels(repeat.dds@colData)
design(repeat.dds) <- ~ PatientNumber + Repeat.Sample

# analysis
repeat.res.list <- list()
repeat.plot.list <- list()

for(i in 1:8){
  cell.type <- levels(repeat.dds$SampleType)[i]
  print(cell.type)
  sub.rep.dds <- repeat.dds[, repeat.dds$SampleType == cell.type]
  sub.rep.dds@colData <- droplevels(sub.rep.dds@colData)
  sub.rep.dds <- DESeq(sub.rep.dds)
  repeat.res <- results(sub.rep.dds, 
                   independentFiltering=TRUE, alpha=0.05)
  repeat.res$Gene <- mcols(repeat.dds)$gene_name
  repeat.res.LFC <- lfcShrink(sub.rep.dds, coef=length(resultsNames(sub.rep.dds)))
  repeat.res$log2FoldChange <- repeat.res.LFC$log2FoldChange
  repeat.res$lfcSE <- repeat.res.LFC$lfcSE
  repeatresOrdered <- repeat.res[order(repeat.res$padj), ]
  repeatresOrdered <- repeatresOrdered[(complete.cases(repeatresOrdered)), ]
  repeatresOrdered$sig <- ifelse((repeatresOrdered$padj < 0.05 & 
                            abs(repeatresOrdered$log2FoldChange) >= log(1.5, 2)), 
                          "Sig DE", "Not Sig")
  repeat.res.list[[i]] <- repeatresOrdered
  summary(repeatresOrdered)
  
  repeat.plot.list[[i]] <- ggplot(data.frame(repeatresOrdered),
                                   aes(log2FoldChange, -log10(padj))) +
    geom_point(aes(col=sig)) +
    scale_color_manual(values=c("grey", "red")) +
    theme_bw() +
    ggtitle(cell.type) +
    geom_text_repel(data=subset(data.frame(repeatresOrdered), sig == "Sig DE"),
                    aes(label=Gene))
}

grid.arrange(grobs=repeat.plot.list, ncol=3)
print(repeat.plot.list[[3]])
print(repeat.plot.list[[4]])
```

The gene that is strongly significant in CD14_MET is shown. It encodes a 
predicted transmembrane protein 

```{r repeat-genes-cd14met, warning=FALSE}
g <- which(mcols(repeat.dds)$gene_name == "TMEM185A")
d <- plotCounts(repeat.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease", "PatientNumber", 
                           "Repeat.Sample"),
                returnData=TRUE)
ggplot(d, aes(x=Repeat.Sample, y=count, color=Disease, group=PatientNumber)) + 
  geom_point(size=3) + 
  #scale_y_log10() +
  theme_bw() +
  geom_line() +
  ylab("TMEM185A") +
  facet_wrap(~ SampleType, nrow=2)
```

```{r repeat-xgr, fig.height=12, message=FALSE}
data <- as.character(repeat.res.list[[4]]$Gene[repeat.res.list[[4]]$sig == "Sig DE"])
background <- as.character(repeat.res.list[[4]]$Gene)
eTerm_repeat_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                      ontology="GOBP",
                                      ontology.algorithm="lea",
                                     RData.location=RData.location,
                                     verbose=TRUE)
bp_repeat_GOBP_lea <- xEnrichBarplot(eTerm_repeat_GOBP_lea,
                                   displayBy="fdr", top_num=10)
bp_repeat_GOBP_lea
```

The effect of metformin treatment was explored by adding an interaction term to 
this analysis

```{r metformin-treatment}
repeat.dds$Metformin. <- as.factor(repeat.dds$Metformin.)
repeat.res.list <- list()
repeat.plot.list <- list()

for(i in 1:8){
  cell.type <- levels(repeat.dds$SampleType)[i]
  print(cell.type)
  sub.rep.dds <- repeat.dds[, repeat.dds$SampleType == cell.type]
  sub.rep.dds@colData <- droplevels(sub.rep.dds@colData)
  
  tmp <- cbind(sub.rep.dds$Metformin., sub.rep.dds$PatientNumber)
  # 10 patients in control, 9 in disease
  # update disease to 1:9 and control 1:10
  ind.n <- tmp[, 2]
  ind.n[which(tmp[, 1] == 1)] <- as.numeric(as.factor(ind.n[which(tmp[, 1] == 1)]))
  ind.n[which(tmp[, 1] == 2)] <- as.numeric(as.factor(ind.n[which(tmp[, 1] == 2)]))

  sub.rep.dds$ind.n <- as.factor(as.character(ind.n))
  design(sub.rep.dds) <- ~ Metformin. + Metformin.:ind.n + Metformin.:Repeat.Sample
  m1 <- model.matrix(design(sub.rep.dds), sub.rep.dds@colData)
  all.zero <- colSums(m1) == 0
  idx <- which(all.zero)
  m1 <- m1[, -idx]  
  
  sub.rep.dds <- DESeq(sub.rep.dds, full=m1, betaPrior=FALSE)
  repeat.res <- results(sub.rep.dds, 
                   independentFiltering=TRUE, alpha=0.05)
  repeat.res$Gene <- mcols(repeat.dds)$gene_name
  repeat.res.LFC <- lfcShrink(sub.rep.dds, coef=length(resultsNames(sub.rep.dds)))
  repeat.res$log2FoldChange <- repeat.res.LFC$log2FoldChange
  repeat.res$lfcSE <- repeat.res.LFC$lfcSE
  repeatresOrdered <- repeat.res[order(repeat.res$padj), ]
  repeatresOrdered <- repeatresOrdered[(complete.cases(repeatresOrdered)), ]
  repeatresOrdered$sig <- ifelse((repeatresOrdered$padj < 0.05 & 
                            abs(repeatresOrdered$log2FoldChange) >= log(1.5, 2)), 
                          "Sig DE", "Not Sig")
  repeat.res.list[[i]] <- repeatresOrdered
  summary(repeatresOrdered)
  
  repeat.plot.list[[i]] <- ggplot(data.frame(repeatresOrdered),
                                   aes(log2FoldChange, -log10(padj))) +
    geom_point(aes(col=sig)) +
    scale_color_manual(values=c("grey", "red")) +
    theme_bw() +
    ggtitle(cell.type) +
    geom_text_repel(data=subset(data.frame(repeatresOrdered), sig == "Sig DE"),
                    aes(label=Gene))
}

grid.arrange(grobs=repeat.plot.list, ncol=3)

g <- which(mcols(repeat.dds)$gene_name == "CTD-2031P19.4")
d <- plotCounts(repeat.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease", "Repeat.Sample", "Metformin.",
                           "PatientNumber"),
                returnData=TRUE)
ggplot(subset(d, SampleType == "CD14_MET"), 
       aes(x=Repeat.Sample, y=count, group=PatientNumber, colour=Disease)) + 
  facet_grid(~ Metformin.) +
  geom_line() +
  geom_point(size=3) + 
  theme_bw() +
  ylab("CTD-2031P19.4")


g <- which(mcols(repeat.dds)$gene_name == "IGLC3")
d <- plotCounts(repeat.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease", "Repeat.Sample", "Metformin.",
                           "PatientNumber"),
                returnData=TRUE)
ggplot(subset(d, SampleType == "Paxgene"), 
       aes(x=Repeat.Sample, y=count, group=PatientNumber, colour=Disease)) + 
  facet_grid(~ Metformin.) +
  geom_line() +
  geom_point(size=3) + 
  theme_bw() +
  ylab("IGLC3")

g <- which(mcols(repeat.dds)$gene_name == "IGLC2")
d <- plotCounts(repeat.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease", "Repeat.Sample", "Metformin.",
                           "PatientNumber"),
                returnData=TRUE)
ggplot(subset(d, SampleType == "Paxgene"), 
       aes(x=Repeat.Sample, y=count, group=PatientNumber, colour=Disease)) + 
  facet_grid(~ Metformin.) +
  geom_line() +
  geom_point(size=3) + 
  theme_bw() +
  ylab("IGLC2")

g <- which(mcols(repeat.dds)$gene_name == "ANK1")
d <- plotCounts(repeat.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease", "Repeat.Sample", "Metformin.",
                           "PatientNumber"),
                returnData=TRUE)
ggplot(subset(d, SampleType == "Paxgene"), 
       aes(x=Repeat.Sample, y=count, group=PatientNumber, colour=Disease)) + 
  facet_grid(~ Metformin.) +
  geom_line() +
  geom_point(size=3) + 
  theme_bw() +
  ylab("ANK1")

g <- which(mcols(repeat.dds)$gene_name == "MZB1")
d <- plotCounts(repeat.dds, gene=g, replaced=TRUE,
                intgroup=c("SampleType", "Disease", "Repeat.Sample", "Metformin.",
                           "PatientNumber"),
                returnData=TRUE)
ggplot(subset(d, SampleType == "Paxgene"), 
       aes(x=Repeat.Sample, y=count, group=PatientNumber, colour=Disease)) + 
  facet_grid(~ Metformin.) +
  geom_line() +
  geom_point(size=3) + 
  theme_bw() +
  ylab("MZB1")
```






### CD14 vs CD8

```{r cd14-cd8, message=FALSE}
MT.res <- results(lps.dds, independentFiltering=TRUE, alpha=0.05,
                  contrast=c("SampleType", "CD14", "CD8"))
MT.res$Gene <- mcols(lps.dds)$gene_name
MT.res.LFC <- lfcShrink(lps.dds, 
                        contrast=c("SampleType", "CD14", "CD8"))
MT.res$log2FoldChange <- MT.res.LFC$log2FoldChange
MT.res$lfcSE <- MT.res.LFC$lfcSE
MTresOrdered <- MT.res[order(MT.res$padj), ]
MTresOrdered <- MTresOrdered[(complete.cases(MTresOrdered)), ]
summary(MTresOrdered)

MTresOrdered$sig <- ifelse((MTresOrdered$padj < 0.05 & 
                            abs(MTresOrdered$log2FoldChange) >= log(1.5, 2)), 
                          "Sig DE", "Not Sig")
ggplot(data.frame(MTresOrdered), aes(log2FoldChange, -log10(padj))) +
  geom_point(aes(col=sig)) +
  scale_color_manual(values=c("grey", "red")) +
  theme_bw() +
  ggtitle("CD8 vs CD14") +
  geom_text_repel(data=subset(data.frame(MTresOrdered), padj <5e-330),
                aes(label=Gene))

r.lfc <- range(MT.res.LFC$log2FoldChange, na.rm=TRUE)
plotMA(MT.res.LFC, ylim=r.lfc, main="MA Plot")
abline(h=c(-log(1.5, 2), log(1.5, 2)), col="dodgerblue", lwd=2)

data <- unique(as.character(MTresOrdered$Gene[MTresOrdered$sig == "Sig DE"]))
background <- unique(as.character(MTresOrdered$Gene))
eTerm_MT_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                        ontology="GOBP",
                                        ontology.algorithm="lea",
                                        RData.location=RData.location, 
                                    verbose=FALSE)
bp_MT_GOBP_lea <- xEnrichBarplot(eTerm_MT_GOBP_lea,
                                   displayBy="fdr", top_num=10)
bp_MT_GOBP_lea
```

### PBMC vs Paxgene

```{r pbmc-paxgene, message=FALSE}
pp.res <- results(lps.dds, independentFiltering=TRUE, alpha=0.05,
                  contrast=c("SampleType", "PBMC", "Paxgene"))
pp.res$Gene <- mcols(lps.dds)$gene_name
pp.res.LFC <- lfcShrink(lps.dds, 
                        contrast=c("SampleType", "PBMC", "Paxgene"))
pp.res$log2FoldChange <- pp.res.LFC$log2FoldChange
pp.res$lfcSE <- pp.res.LFC$lfcSE
ppresOrdered <- pp.res[order(pp.res$padj), ]
ppresOrdered <- ppresOrdered[(complete.cases(ppresOrdered)), ]
summary(ppresOrdered)

ppresOrdered$sig <- ifelse((ppresOrdered$padj < 0.05 & 
                            abs(ppresOrdered$log2FoldChange) >= log(1.5, 2)), 
                          "Sig DE", "Not Sig")
ggplot(data.frame(ppresOrdered), aes(log2FoldChange, -log10(padj))) +
  
  geom_point(aes(col=sig)) +
  scale_color_manual(values=c("grey", "red")) +
  theme_bw() +
  ggtitle("PBMC vs Paxgene") +
  geom_text_repel(data=subset(data.frame(ppresOrdered), padj <5e-250),
                aes(label=Gene))

data <- unique(as.character(ppresOrdered$Gene[ppresOrdered$sig == "Sig DE"]))
background <- unique(as.character(ppresOrdered$Gene))
eTerm_pp_GOBP_lea <- xEnricherGenes(data=data, background=background,
                                        ontology="GOBP",
                                        ontology.algorithm="lea",
                                        RData.location=RData.location, 
                                    verbose=FALSE)
bp_pp_GOBP_lea <- xEnrichBarplot(eTerm_pp_GOBP_lea,
                                   displayBy="fdr", top_num=10)
bp_pp_GOBP_lea
```

### Quantitative traits

Looking at BMI in each cell type finds no differentially expressed genes. 

```{r bmi, message=FALSE}
# design
bmi.dds <- lps.dds
# 152 samples remaining (19 individuals, 8 samples each)
design(bmi.dds) <- ~ BMI

# analysis
bmi.res.list <- list()
bmi.plot.list <- list()

for(i in 1:8){
  cell.type <- levels(bmi.dds$SampleType)[i]
  print(cell.type)
  sub.bmi.dds <- bmi.dds[, bmi.dds$SampleType == cell.type]
  sub.bmi.dds@colData <- droplevels(sub.bmi.dds@colData)
  sub.bmi.dds <- DESeq(sub.bmi.dds)
  bmi.res <- results(sub.bmi.dds, 
                   independentFiltering=TRUE, alpha=0.05)
  bmi.res$Gene <- mcols(bmi.dds)$gene_name
  bmi.res.LFC <- lfcShrink(sub.bmi.dds, coef=length(resultsNames(sub.bmi.dds)))
  bmi.res$log2FoldChange <- bmi.res.LFC$log2FoldChange
  bmi.res$lfcSE <- bmi.res.LFC$lfcSE
  bmiresOrdered <- bmi.res[order(bmi.res$padj), ]
  bmiresOrdered <- bmiresOrdered[(complete.cases(bmiresOrdered)), ]
  bmi.res.list[[i]] <- bmiresOrdered
  summary(bmiresOrdered)
  
  bmiresOrdered$sig <- ifelse((bmiresOrdered$padj < 0.05), 
                          "Sig DE", "Not Sig")
  bmi.plot.list[[i]] <- ggplot(data.frame(bmiresOrdered),
                                   aes(log2FoldChange, -log10(padj))) +
    geom_point(aes(col=sig)) +
    scale_color_manual(values=c("grey", "red")) +
    theme_bw() +
    ggtitle(cell.type) +
    geom_text_repel(data=subset(data.frame(bmiresOrdered), sig == "Sig DE"),
                    aes(label=Gene))
}

grid.arrange(grobs=bmi.plot.list, ncol=3)
plot(bmi.plot.list[[6]])
```

```{r bmi-disease, message=FALSE}
# design
bmi.dds <- lps.dds
# 152 samples remaining (19 individuals, 8 samples each)
design(bmi.dds) <- ~ Disease + BMI

# analysis
bmi.res.list <- list()
bmi.plot.list <- list()

for(i in 1:8){
  cell.type <- levels(bmi.dds$SampleType)[i]
  print(cell.type)
  sub.bmi.dds <- bmi.dds[, bmi.dds$SampleType == cell.type]
  sub.bmi.dds@colData <- droplevels(sub.bmi.dds@colData)
  sub.bmi.dds <- DESeq(sub.bmi.dds)
  bmi.res <- results(sub.bmi.dds, 
                   independentFiltering=TRUE, alpha=0.05)
  bmi.res$Gene <- mcols(bmi.dds)$gene_name
  bmi.res.LFC <- lfcShrink(sub.bmi.dds, coef=length(resultsNames(sub.bmi.dds)))
  bmi.res$log2FoldChange <- bmi.res.LFC$log2FoldChange
  bmi.res$lfcSE <- bmi.res.LFC$lfcSE
  bmiresOrdered <- bmi.res[order(bmi.res$padj), ]
  bmiresOrdered <- bmiresOrdered[(complete.cases(bmiresOrdered)), ]
  bmi.res.list[[i]] <- bmiresOrdered
  summary(bmiresOrdered)
  
  bmiresOrdered$sig <- ifelse((bmiresOrdered$padj < 0.05), 
                          "Sig DE", "Not Sig")
  bmi.plot.list[[i]] <- ggplot(data.frame(bmiresOrdered),
                                   aes(log2FoldChange, -log10(padj))) +
    geom_point(aes(col=sig)) +
    scale_color_manual(values=c("grey", "red")) +
    theme_bw() +
    ggtitle(cell.type) +
    geom_text_repel(data=subset(data.frame(bmiresOrdered), sig == "Sig DE"),
                    aes(label=Gene))
}

grid.arrange(grobs=bmi.plot.list, ncol=3)
```






```{r glucose, message=FALSE}
# design
glu.dds <- lps.dds
# 152 samples remaining (19 individuals, 8 samples each)
design(glu.dds) <- ~ Glucose..mmol.l..Consent..FG.sheet.

# analysis
glu.res.list <- list()
glu.plot.list <- list()

for(i in 1:8){
  cell.type <- levels(glu.dds$SampleType)[i]
  print(cell.type)
  sub.glu.dds <- glu.dds[, glu.dds$SampleType == cell.type]
  sub.glu.dds@colData <- droplevels(sub.glu.dds@colData)
  sub.glu.dds <- DESeq(sub.glu.dds)
  glu.res <- results(sub.glu.dds, 
                   independentFiltering=TRUE, alpha=0.05)
  glu.res$Gene <- mcols(glu.dds)$gene_name
  glu.res.LFC <- lfcShrink(sub.glu.dds, coef=length(resultsNames(sub.glu.dds)))
  glu.res$log2FoldChange <- glu.res.LFC$log2FoldChange
  glu.res$lfcSE <- glu.res.LFC$lfcSE
  gluresOrdered <- glu.res[order(glu.res$padj), ]
  gluresOrdered <- gluresOrdered[(complete.cases(gluresOrdered)), ]
  glu.res.list[[i]] <- gluresOrdered
  summary(gluresOrdered)
  
  gluresOrdered$sig <- ifelse((gluresOrdered$padj < 0.05), 
                          "Sig DE", "Not Sig")
  glu.plot.list[[i]] <- ggplot(data.frame(gluresOrdered),
                                   aes(log2FoldChange, -log10(padj))) +
    geom_point(aes(col=sig)) +
    scale_color_manual(values=c("grey", "red")) +
    theme_bw() +
    ggtitle(cell.type) +
    geom_text_repel(data=subset(data.frame(gluresOrdered), sig == "Sig DE"),
                    aes(label=Gene))
}

grid.arrange(grobs=glu.plot.list, ncol=3)
plot(glu.plot.list[[1]])
plot(glu.plot.list[[2]])
plot(glu.plot.list[[3]])
plot(glu.plot.list[[4]])
plot(glu.plot.list[[5]])
plot(glu.plot.list[[6]])
plot(glu.plot.list[[7]])
plot(glu.plot.list[[8]])
```