---
title: "Abu Dhabi RNA-seq project: analysis of paxgene data set"
author: "Katie Burnham"
date: '`r Sys.Date()`'
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE)

library(kburnhamfunctions)
# install.packages("Vennerable", repos = "http://R-Forge.R-project.org")

# Vennerable needs graph and RBGL, from bioconductor repository
# source("http://bioconductor.org/biocLite.R")
# biocLite("graph")
# biocLite("RBGL")

# sample size
library(ssizeRNA)

# Those packages also need reshape and gtools
library(reshape)
library(gtools)
library(Vennerable)

# Differential expression analysis
library(DESeq2)

# plotting figures
library(dplyr)
library(ggplot2)
library(ggrepel)
library(gridExtra)

# normalisation
library(vsn)

# heatmap
library(pheatmap)
library(RColorBrewer)

# parallel computing
library(BiocParallel)
register(SnowParam(4))

# look up gene ids
library(ensembldb)
library(EnsDb.Hsapiens.v75)

# pathway analysis
library(XGR)
```

```{r load-data}
load("PaxGeneCohort/dds.Rdata")
load("U:/PANDIT/transcript_metadata.RData")
colnames(attributeData)[1] <- "chr"
m1 <- match(rownames(assay(dds)), attributeData$gene_id)
assay.info <- attributeData[m1, ]
mcols(dds) <- assay.info
table(mcols(dds)$gene_biotype)
```

```{r lps-response, message=FALSE}
dds$Sex <- as.factor(dds$Sex)
design(dds) <- ~ BMI + Sex + Control.or.T2D
dds <- dds[, !is.na(dds@colData$BMI)]
dds <- dds[, !is.na(dds@colData$Sex)]
dds
dds@colData <- droplevels(dds@colData)

# analysis
dds <- DESeq(dds, parallel=TRUE)
save(dds, file="PaxGeneCohort/dds-deseq.Rdata")

# results
dds.res <- results(dds, parallel=TRUE,
                   contrast=c("Control.or.T2D", "T2D", "Control"),
                   independentFiltering=TRUE, alpha=0.05)
dds.res$Gene <- mcols(dds)$gene_name

plot(metadata(dds.res)$filterNumRej,
     type="b", ylab="Number of rejections",
     xlab="Quantiles of filter", pch=16)
lines(metadata(dds.res)$lo.fit, col="red")
abline(v=metadata(dds.res)$filterTheta)

dds.res.LFC <- lfcShrink(dds, 
                         contrast=c("Control.or.T2D", "T2D", "Control"),
                         parallel=TRUE)
lps.res$log2FoldChange <- lps.res.LFC$log2FoldChange
lps.res$lfcSE <- lps.res.LFC$lfcSE
lpsresOrdered <- lps.res[order(lps.res$padj), ]
lpsresOrdered <- lpsresOrdered[(complete.cases(lpsresOrdered)), ]
summary(lpsresOrdered)
res05 <- results(lps.dds, alpha=0.05)

# write.table(lpsresOrdered,
            # file="U:/Abu-Dhabi/RNASeq/PilotData/LPSResponse_All_Patients.txt",
            # sep="\t")
```

